proof/refine/AARCH64/orphanage/Orphanage.thy

```diff
@@ -545,60 +503,75 @@ lemma activateThread_no_orphans [wp]:
+   threadSet (tcbQueued_update (\<lambda>_. False)) tcb_ptr
+   \<lbrace>\<lambda>_ s. tcb_ptr' \<in> all_queued_tcb_ptrs s\<rbrace>"
+  apply (wpsimp wp: threadSet_wp)
+  apply (clarsimp simp: all_queued_tcb_ptrs_def obj_at'_def ps_clear_upd)
   done
 
-lemma setQueue_almost_no_orphans_deq [wp]:
-  "\<lbrace> \<lambda>s. almost_no_orphans tcb_ptr s \<and>
-        queue = [x\<leftarrow>((ksReadyQueues s) (d, priority)). x \<noteq> tcb_ptr] \<rbrace>
-   setQueue d priority queue
-   \<lbrace> \<lambda>rv s. almost_no_orphans tcb_ptr s \<rbrace>"
-  unfolding setQueue_def
-  apply (wp | clarsimp)+
-  apply (fastforce simp: almost_no_orphans_def all_queued_tcb_ptrs_def
-                        all_active_tcb_ptrs_def is_active_tcb_ptr_def)
+lemma tcbSchedDequeue_all_queued_tcb_ptrs_other:
+  "\<lbrace>\<lambda>s. tcb_ptr \<noteq> tcb_ptr' \<and> tcb_ptr' \<in> all_queued_tcb_ptrs s\<rbrace>
+   tcbSchedDequeue tcb_ptr
+   \<lbrace>\<lambda>_ s. tcb_ptr' \<in> all_queued_tcb_ptrs s\<rbrace>"
+  unfolding tcbSchedDequeue_def
+  by (wpsimp wp: tcbQueued_update_False_all_queued_tcb_ptrs threadGet_wp)
+
+lemma tcbQueued_update_False_almost_no_orphans:
+  "\<lbrace>no_orphans\<rbrace>
+   threadSet (tcbQueued_update (\<lambda>_. False)) tptr
+   \<lbrace>\<lambda>_. almost_no_orphans tptr\<rbrace>"
+  apply (wpsimp wp: threadSet_wp)
+  apply (clarsimp simp: no_orphans_def almost_no_orphans_def)
+  apply (rename_tac tcb_ptr)
+  apply (case_tac "tcb_ptr = tptr")
+   apply fastforce
+  apply (fastforce simp: all_queued_tcb_ptrs_def obj_at'_def all_active_tcb_ptrs_def
+                         is_active_tcb_ptr_def st_tcb_at'_def ps_clear_upd)
   done
 
 lemma tcbSchedDequeue_almost_no_orphans [wp]:
-  "\<lbrace> \<lambda>s. no_orphans s \<rbrace>
-   tcbSchedDequeue thread
-   \<lbrace> \<lambda>rv s. almost_no_orphans thread s \<rbrace>"
+  "\<lbrace>no_orphans\<rbrace> tcbSchedDequeue thread \<lbrace>\<lambda>_. almost_no_orphans thread\<rbrace>"
   unfolding tcbSchedDequeue_def
-  apply (wp threadSet_almost_no_orphans | simp cong: if_cong)+
-  apply (simp add:no_orphans_strg_almost cong: if_cong)
+  apply (wpsimp wp: tcbQueued_update_False_almost_no_orphans threadGet_wp)
+  apply (simp add: no_orphans_strg_almost)
   done
 
-lemma tcbSchedDequeue_no_orphans [wp]:
-  "\<lbrace> \<lambda>s. no_orphans s \<and> \<not> is_active_tcb_ptr tcb_ptr s \<rbrace>
-   tcbSchedDequeue tcb_ptr
-   \<lbrace> \<lambda>rv s. no_orphans s \<rbrace>"
-  unfolding tcbSchedDequeue_def
-  apply (wp setQueue_no_orphans_deq threadSet_no_orphans | clarsimp)+
-   apply (wp getObject_tcb_wp | clarsimp simp: threadGet_def)+
-  apply (drule obj_at_ko_at')
-  apply auto
+lemma tcbSchedDequeue_no_orphans[wp]:
+  "\<lbrace>\<lambda>s. no_orphans s \<and> \<not> is_active_tcb_ptr tcbPtr s \<and> tcb_at' tcbPtr s\<rbrace>
+   tcbSchedDequeue tcbPtr
+   \<lbrace>\<lambda>_. no_orphans\<rbrace>"
+  supply disj_not1[simp del]
+  apply (unfold no_orphans_disj almost_no_orphans_disj)
```

Comment:
why not `unfolding ...` and then the `supply`?

Reply:
I sort of thought `supply` bits came first, but I can change this to whatever. Some of these Orphanage proofs might have a slightly older flavour since I was reworking old proofs

Reply:
Yeah, good point. The `supply` can stay first. So maybe only tweak the `apply` into an `unfolding`.